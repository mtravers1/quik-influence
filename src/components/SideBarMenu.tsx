import { Flex, Box } from '@chakra-ui/layout';
import React, { useEffect, useState } from 'react';
import quikColorConstants, { sidebarBg } from 'utils/constants/colorConstants';
import NextLink from './NextLink';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { css } from '@emotion/react';
import { SideBarMenuOptions } from 'modules';
import { SideBarOptionMenu } from 'types';
import { useRouter } from 'next/router';
import {
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
  ColorMode,
} from '@chakra-ui/react';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

interface SideBarMenuProps {
  bgColor?: string;
  color?: string;
  colorMode: ColorMode;
  open: boolean;
}

const SideBarMenu = ({ bgColor, colorMode, open }: SideBarMenuProps) => {
  const route = useRouter();
  const { pathname, asPath } = route || { pathname: '/' };
  const _sideBarOptions = Object.values(SideBarMenuOptions);

  const navcss = (isActive: boolean) => css`
    & {
      &:hover {
        &::before {
          background: ${quikColorConstants.influenceRed};
          height: 80%;
        }
      }

      &::before {
        background: ${isActive
          ? quikColorConstants.influenceRed
          : sidebarBg[colorMode]};
        height: ${isActive ? '80%;' : '0'};
      }
    }
  `;

  const maincss = css`
    & {
      left: ${open ? '0' : '-300px'};
      transition: 0.3s ease;
    }
  `;

  console.log(asPath);

  let currentRoute: string;
  const paths = pathname.split('/');

  if (paths.length === 2) {
    currentRoute = '/dashboard';
  } else currentRoute = paths[2];

  return (
    <Flex
      flexDirection="column"
      width="250px"
      py={10}
      bgColor={bgColor}
      position={{ base: 'fixed', md: 'sticky' }}
      top="60px"
      overflow="hidden"
      flexShrink={0}
      css={maincss}
      height="95vh"
    >
      {_sideBarOptions.map(
        ({ name, icon, path, isShown, child }: SideBarOptionMenu) => {
          if (!isShown) return;

          const currentPath = path.split('/dashboard/')[1] || '/dashboard';
          const isActive = currentRoute === currentPath;

          return (
            <NavComponent
              key={name}
              isActive={isActive}
              colorMode={colorMode}
              navcss={navcss}
              path={path}
              icon={icon}
              child={child}
              asPath={asPath}
              name={name}
            />
          );
        }
      )}
    </Flex>
  );
};

const NavComponent = ({
  isActive,
  colorMode,
  navcss,
  path,
  icon,
  child,
  asPath,
  name,
}: {
  isActive: boolean;
  colorMode: ColorMode;
  navcss: any;
  path: string;
  icon: any;
  child: any;
  asPath: any;
  name: string;
}) => {
  return (
    <>
      <Box
        minW="100%"
        position="relative"
        css={navcss(isActive)}
        _before={{
          content: '""',
          position: 'absolute',
          left: 0,
          top: '50%',
          transform: 'translateY(-50%)',
          height: 0,
          width: '5px',
          background: quikColorConstants.influenceRed,
          borderRadius: '0 10px 10px 0',
          transition: '0.3s ease',
        }}
        bg={isActive ? sidebarBg[colorMode] : 'transparent'}
      >
        {!child.length && (
          <NextLink
            href={path as string}
            display="block"
            _hover={{
              textDecoration: 'none',
              color: 'inherit',
            }}
            _focus={{
              border: 'none',
              textDecoration: 'none',
            }}
            fontFamily="Avenir"
            fontWeight="bold"
            fontSize="16px"
            py={5}
            px={10}
          >
            <FontAwesomeIcon
              icon={icon as IconProp}
              style={{ marginRight: '20px' }}
            />
            {name}
          </NextLink>
        )}
      </Box>

      {!!child?.length && (
        <Accordion allowMultiple paddingLeft="35px">
          <AccordionItem>
            <AccordionButton>{name}</AccordionButton>
            <AccordionPanel pb={4}>
              {child.map((childEl: any) => (
                <NextLink
                  href={childEl.path as string}
                  color={asPath === childEl.path ? 'red' : undefined}
                  display="block"
                  _hover={{
                    textDecoration: 'none',
                    color: 'inherit',
                  }}
                  _focus={{
                    border: 'none',
                    textDecoration: 'none',
                  }}
                  fontFamily="Avenir"
                  fontWeight="bold"
                  py={3}
                  px={10}
                  fontSize="14px"
                >
                  {childEl.name}
                </NextLink>
              ))}
            </AccordionPanel>
          </AccordionItem>
        </Accordion>
      )}
    </>
  );
};

export default SideBarMenu;
