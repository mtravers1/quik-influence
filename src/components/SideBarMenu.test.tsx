import { render } from '@testing-library/react';
import SideBarMenu from './SideBarMenu';

it('renders SideBarMenu component', () => {
  const { container } = render(<SideBarMenu colorMode="dark" open={true} />);
  expect(container).toMatchSnapshot();
});
