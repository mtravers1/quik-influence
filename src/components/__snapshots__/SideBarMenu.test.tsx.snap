// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SideBarMenu component 1`] = `
<div>
  <div
    class="css-vr7qz"
  >
    <div
      class="css-1ix7n7v"
    >
      <a
        class="chakra-link css-qffdt"
        href="/dashboard"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-grip-horizontal "
          data-icon="grip-horizontal"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-right: 20px;"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 288H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm160 0h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm160 0h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zM96 96H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm160 0h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm160 0h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32z"
            fill="currentColor"
          />
        </svg>
        Overview
      </a>
    </div>
    <div
      class="css-63s2m5"
    >
      <a
        class="chakra-link css-qffdt"
        href="/dashboard/campaigns"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-calendar-minus "
          data-icon="calendar-minus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-right: 20px;"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M436 160H12c-6.6 0-12-5.4-12-12v-36c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48v36c0 6.6-5.4 12-12 12zM12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm304 192c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12H132c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h184z"
            fill="currentColor"
          />
        </svg>
        Campaigns
      </a>
    </div>
    <div
      class="css-63s2m5"
    >
      <a
        class="chakra-link css-qffdt"
        href="/dashboard/credits"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-credit-card "
          data-icon="credit-card"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-right: 20px;"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 432c0 26.5 21.5 48 48 48h480c26.5 0 48-21.5 48-48V256H0v176zm192-68c0-6.6 5.4-12 12-12h136c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H204c-6.6 0-12-5.4-12-12v-40zm-128 0c0-6.6 5.4-12 12-12h72c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM576 80v48H0V80c0-26.5 21.5-48 48-48h480c26.5 0 48 21.5 48 48z"
            fill="currentColor"
          />
        </svg>
        Credits
      </a>
    </div>
    <div
      class="css-63s2m5"
    >
      <a
        class="chakra-link css-qffdt"
        href="/dashboard/leads"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user-plus "
          data-icon="user-plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="margin-right: 20px;"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            fill="currentColor"
          />
        </svg>
        Leads
      </a>
    </div>
  </div>
</div>
`;
